### PR TITLE
Added `Id` property to `Identifier`

### LINKED ISSUE #151: Support function effects outside of expression
I want create a function to check a value whether large than 10 three times. 

The Expression will be `Repeat([value] > 10, 3)`.

```
 var e = new Expression("Repeat([value] > 10, 3)");
 
 var times = new Dictionary<string, int>() { };

 e.EvaluateFunction += (name, args, id) =>
 {
     if (name == "Repeat")
     {
         var t = (int)args.Parameters[1].Evaluate() - 1;
         var r = (bool)args.Parameters[0].Evaluate();
         if (r && id != null)
         {
             if (!times.ContainsKey(id))
             {
                 times[id] = t;
             }
             else
             {
                 times[id]--;
             }
         }

         args.Result =  r && times[id] == 0;
     }
 };
 e.Parameters["value"] = 9;

 Assert.Equal(false, e.Evaluate());
 
 e.Parameters["value"] = 11;
 Assert.Equal(false, e.Evaluate());
 e.Parameters["value"] = 12;
 Assert.Equal(false, e.Evaluate()); 
 e.Parameters["value"] = 13;
 Assert.Equal(true, e.Evaluate()); 
```

### PR DESCRIPTION
**Pull request by gumbarros**
This closes #151

### CODE REVIEW

**Comment by gumbarros** [comment]
@sklose check the unit tests, this will solve your case right? Thanks for the original implementation

**Comment by Bykiev** [comment]
> @sklose check the unit tests, this will solve your case right? Thanks for the original implementation

did you mean @slobber?

**Review by Bykiev — APPROVED** [review]
LGTM

File: src/NCalc.Sync/ExpressionFunctionData.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+namespace NCalc;
+
+public class ExpressionFunctionData(Guid id, Expression[] arguments, ExpressionContext context)
```
Comment (gmcchessney) [inline src/NCalc.Sync/ExpressionFunctionData.cs:3]: I don't see any way to get the number of arguments, since `Arguments` is private.

**Pull request merged**